File: Application/Services/LocationService.cs
using Core.DTOs.Location;
using Core.Interfaces.Repositories;
using Core.Interfaces.Services;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Application.Services
{
    public class LocationService : ILocationService
    {
        private readonly ILocationRepository _repo;
        public LocationService(ILocationRepository repo) => _repo = repo;

        public async Task<IEnumerable<DistrictDto>> GetAllDistrictsAsync()
        {
            var districts = await _repo.GetAllDistrictsAsync();
            return districts.Select(d => new DistrictDto
            {
                Id = d.Id,
                Name = d.Name,
                Code = d.Code
            });
        }

        public async Task<IEnumerable<ConstituencyDto>> GetConstituenciesByDistrictAsync(int districtId)
        {
            var items = await _repo.GetConstituenciesByDistrictAsync(districtId);
            return items.Select(x => new ConstituencyDto
            {
                Id = x.Id,
                Name = x.Name,
                Code = x.Code,
                DistrictId = x.DistrictId
            });
        }

        public async Task<IEnumerable<AreaDto>> GetAreasByConstituencyAsync(int constituencyId)
        {
            var items = await _repo.GetAreasByConstituencyAsync(constituencyId);
            return items.Select(x => new AreaDto
            {
                Id = x.Id,
                Name = x.Name,
                Code = x.Code,
                ConstituencyId = x.ConstituencyId
            });
        }

        public async Task<IEnumerable<BlockDto>> GetBlocksByAreaAsync(int areaId)
        {
            var items = await _repo.GetBlocksByAreaAsync(areaId);
            return items.Select(x => new BlockDto
            {
                Id = x.Id,
                Name = x.Name,
                Code = x.Code,
                AreaId = x.AreaId
            });
        }
    }
}

File: citizen_connect_python/app/services/location_service.py
from app.repositories.location_repository import LocationRepository
from app.schemas.location import (
    DistrictSchema, ConstituencySchema, AreaSchema, BlockSchema
)


class LocationService:
    def __init__(self, repo: LocationRepository):
        self.repo = repo

    async def get_all_districts(self) -> list[DistrictSchema]:
        districts = await self.repo.get_all_districts()
        return [DistrictSchema.model_validate(d) for d in districts]

    async def get_constituencies_by_district(
        self, district_id: int
    ) -> list[ConstituencySchema]:
        items = await self.repo.get_constituencies_by_district(district_id)
        return [ConstituencySchema.model_validate(i) for i in items]

    async def get_areas_by_constituency(
        self, constituency_id: int
    ) -> list[AreaSchema]:
        items = await self.repo.get_areas_by_constituency(constituency_id)
        return [AreaSchema.model_validate(i) for i in items]

    async def get_blocks_by_area(self, area_id: int) -> list[BlockSchema]:
        items = await self.repo.get_blocks_by_area(area_id)
        return [BlockSchema.model_validate(i) for i in items]